flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/SnapshotObserver.java

```diff
@@ -74,8 +80,32 @@ public void observeSavepointStatus(FlinkResourceContext<CR> ctx) {
         cleanupSavepointHistory(ctx, savepointInfo);
     }
 
+    public void observeCheckpointStatus(FlinkResourceContext<CR> ctx) {
+        if (!isCheckpointsTriggeringSupported(ctx.getObserveConfig())) {
+            return;
+        }
+        var resource = ctx.getResource();
+        var jobStatus = resource.getStatus().getJobStatus();
+        var checkpointInfo = jobStatus.getSavepointInfo();
+        var jobId = jobStatus.getJobId();
+
+        // If any manual or periodic savepoint is in progress, observe it
+        if (SnapshotUtils.checkpointInProgress(jobStatus)) {
+            observeTriggeredCheckpoint(ctx, jobId);
+        }
+
+        // REVIEW: clarify if this is relevant for checkpoints.
+        /*
+        // If job is in globally terminal state, observe last savepoint
+        if (ReconciliationUtils.isJobInTerminalState(resource.getStatus())) {
+            observeLatestCheckpoint(
+                    ctx.getFlinkService(), checkpointInfo, jobId, ctx.getObserveConfig());
+        }
+        */
```

Comment:
I think it is relevant. When the job terminates, we might still have completed the externally triggered checkpoint.

Reply:
@gyfora there is something I do not quite get in the current logic: the symmetric `observeLatestSavepoint` uses `getLastCheckpoint` that does not explicitly make a distinction between savepoints and checkpoints. IIUC, if a checkpoint is stored to an externally-addressable location, it will be returned and potentially used for recovery. Is that desirable? I guess in this case we do not need to necessarily need to introduce `observeLatestCheckpoint `.

Reply:
The point of this logic is not necessarily to return the latest savepoint but instead to return the latest state for upgrade purposes (checkpoint or savepoint). Since terminal jobs won't take any more we know it is the last one and we can perform the upgrade.

So even if the upgradeMode is last-state / savepoint, for terminal jobs we simply observe the latest checkpoint info from the REST API (it is not present anywhere else because HA metadata is gone.)

Reply:
So, it seems that `observeLatestSavepoint` therefore already covers all of the lifecycle and recovery needs and we do not necessarily need to introduce the `observeLatestCheckpoint ` method, am I right?

Reply:
yes, I believe so

Reply:
Thanks, I removed that bit.